### PR TITLE
#135: add noop

### PR DESCRIPTION
**Pull request by leaumar**
implement #135

### CODE REVIEW

File: test/index.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ import {
   Merge,
   NonEmptyObject,
   NonNever,
+  noop,
+  noop as erase,
```
Comment (macbem) [inline test/index.ts]: This line is unnecessary - you can just use the previous import.

Comment (quezak) [reply test/index.ts]: yes, also the name `erase` is misleading, as it suggests that the function _does something_

File: test/index.ts [hunk 2]
```diff
@@ -447,3 +449,10 @@ function testXOR() {
   type Test1 = Assert<IsExact<Actual1, Expected1>>;
   type Test2 = Assert<IsExact<Actual2, Expected2>>;
 }
+
+function testNoop() {
+  const x: void = noop();
+  const y: void = noop(false, 0, "", {}, [], null, undefined, Promise.resolve(), new Error(), noop);
+  const z: Promise<void> = Promise.resolve("foo").then(erase);
+  const callback: (_: number) => void = Math.random() < 0.5 ? (i => void i) : noop;
```
Comment (quezak) [inline test/index.ts]: the `void` operator is weird ;)
```suggestion
  const callback: (_: number) => void = Math.random() < 0.5 ? (() => {}) : noop;
```

File: test/index.ts (hunk 1 above)
Comment (leaumar) [reply test/index.ts]: I guess this file isn't used as any form of user documentation, so providing a semantic example use isn't meaningful, you're right

File: test/index.ts (hunk 2 above)
Comment (leaumar) [reply test/index.ts]: I always find `{}` dubious since it looks like an empty code block or an object literal at the same time, but sure

Comment (leaumar) [reply test/index.ts]: Also, I guess if this file isn't meant to serve as documentation, the `i` callback for assignment equality showcasing isn't useful

**Comment by leaumar** [comment]
I did run `test:fix` several times, but then that one last little edit I did without running it got me :)

File: README.md [hunk 3]
```diff
@@ -58,6 +58,7 @@ npm install --save-dev ts-essentials
 ### Basic
 
 - `Primitive` type matching all primitive values.
+- `noop` function that takes any arguments and returns nothing. Makes a convenient default callback, Promise value eraser, etc.
```
Comment (leaumar) [inline README.md]: should this wording perhaps avoid the use of "etc" and just be 2 examples without leading on?

Comment (macbem) [reply README.md]: TBH I'd avoid listing `Promise value eraser` as an example - you're the first person I've seen using this naming. A `noop` is a `noop` - a function that accepts anything, does nothing and returns nothing, just a convenient placeholder fn.

**Comment by krzkaczor** [comment]
Hey, sorry to jump on this so late, after the main discussion. Two things:

1. Have we discussed returning `any` instead of `void`? This way it could be used as a dummy implementation for anything (a littel bit different than simple noop).

I thought that we could combine this with https://github.com/krzkaczor/ts-essentials/issues/134 

Basically we could have: 
```
export const AnyValue = undefined as any;
export const AnyFunc = (): any => { }
```
`AnyFunc` would be (more or less) the same as `noop` proposed in this PR.

2. Do we need: `..._args: unknown[]` at all? isn't empty arg list would work the same?

CC: @quezak

**Comment by quezak** [comment]
1. `() => void` and () => any` is a different noop. I'd say both are useful in different scenarios, so now I don't know which one should be named just `noop` :sweat_smile: Though I liked the `void` one better, what are the cases where you can't use it and can use the `any` noop?

// And if we're adding "different noops", there's also the "identity noop mapping"  `const id = <T>(x: T) => T` :D

Your proposals are OK, though I'd export them as `ANY_VALUE` (it's a constant, not a type/class/etc) and "some kind of noop" (definitely not capitalized, it's not a type).

2. Yeah, explicit args shouldn't be required. @leaumar did you have any usecase that doesn't work if the function is just `() => {}`?

**Comment by leaumar** [comment]
Some thoughts of mine:
> returning `any` instead of `void`

`noop-ts` does this, and IMO the `any` keyword should be avoided as much as possible because it can sneakily introduce bugs by removing type checking.

> Do we need `..._args`

I _thought_ we did, but I just tried and it seems typescript allows you to assign a parameterless function to a variable typed as a function _with_ parameters, so for that part I guess we don't.

You'd still need it if you want to invoke the function explicitly with some parameters though, e.g. to swallow an error. Now I know it was mentioned before that that comes down to circumventing linters, but I don't entirely agree with that. It is in some coding styles and project configs, but it isn't in others. I for one find `catch (expected) { noop(expected); }` for example a nicely explicit, short, refactor-proof pattern that isn't dependent on configuring any code inspections tools to let it pass.
I don't see any downsides to letting it accept arguments anyway, so we may as well in order to cover more bases, no?

> And if we're adding "different noops"

Perhaps instead of following the simple old `noop` trope, `ts-essentials` could bring something more versatile, like this? (Names just quick illustrative spitballing)
```typescript
// idle, inert, dummy, placeholder, something like that
export const noble = {
  empty = () => {},
  consume = (..._args) => {},
  identity = function<T>(x: T) {return xl},
  // more variations
};
```

**Comment by quezak** [comment]
> You'd still need it if you want to invoke the function explicitly with some parameters 

yeah, for me that's an arugment to keep the unused args there

> `catch (expected) { noop(expected); } `

For me this is not explicit enough ;) but it would be nice with `import { noop as ignoreResult }` or something.

I think let's for now add a plain `noop` as propsed originally, returning void with ignored args, and think later what more is needed.

**Comment by leaumar** [comment]
I guess for the future, if functions were to be added and grouped into an overarching object, setting up a double export like `export const noop = dummy.fn` for backward compatibility wouldn't be too bad.

**Comment by krzkaczor** [comment]
@leaumar @quezak I love this discussion, thanks!

> I think let's for now add a plain noop as propsed originally, returning void with ignored args, and think later what more is needed.

AGREED. Let's just merge it as a "sane" noop. We will continue with what I proposed on the appropriate issue.

**Comment by krzkaczor** [comment]
@all-contributors please add @leaumar for code, ideas, doc

**Comment by krzkaczor** [comment]
Thanks again @leaumar !

Published as `ts-essentials@6.0.3`

**Comment by leaumar** [comment]
Thanks for merging :) Will be using it right away

**Pull request merged**